support multi character _record_delimiter during load

Astralidea opened the pull request:
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #3857

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->
There are some requirements that need to distinguish the data of ```<br>```
at this time, we need to support multi character record delimiter

trueeyu committed:
AssertNodes should not output const column (#3820)

All ExecNodes should not direct return const column,
otherwise all exec nodes should check if it have const column.

Current most get_next interface() in ExecNodes can't process the const column of input chunk.

Pslydhh committed:
Add array_intersect (#3652)

HangyuanLiu committed:
Support TEMPORARY partition and select into outfile in new Parser (#3736)

Astralidea committed:
support multi character _record_delimiter during load

Astralidea committed:
Fix.

Astralidea committed:
Fix.

stdpain committed:
Fix some NPE in FE when collect global dictionary (#3824)

stdpain committed:
Fix chunk size cause crash when rolling upgrade (#3835)

murphyatwork committed:
Benchmark sorter (#3790)

trueeyu committed:
Remove unused log (#3859)

silverbullet233 committed:
Fix predicates pushdown in external tables (#3803)

trueeyu committed:
add memory statistics for tablet meta which load from disk (#3850)

qinpengxiang001 committed:
Fix dead lock in FE when collect global dictionary (#3836)

satanson committed:
fixup: disorder of GRF components caused by wrong order of FragmentInstance'backendId (#3865)

karan-kap00r committed:
Replace boost::variant with std::variant  (#3815)

decster committed:
Bugfix: change apply thread name to be less than 16 bytes (#3861)

Apply thread was too long and set_thread_name method will print warning log about this, this PR change the name to be less than 16 bytes.

HangyuanLiu committed:
Support sql mode in new Parser (#3856)

sduzh committed:
Make APIs of RandomAccessFile similar to those of io::RandomAccessFile (#3565)

- Redefine `Status RandomAccessFile::read(Slice* result)` as `StatusOr<int64_t> RandomAccessFile::read(void* data, int64_t size)`;
- Redefine `Status RandomAccessFile::read_at(uint64_t off, const Slice& result)` as `StatusOr RandomAccessFile::read_at_fully(int64_t off, void* data, int64_t size)`;
- Redefine `Status SequentialFile::read(Slice* res)` as `StatusOr<int64_t> SequentialFile::read(void* data, int64_t size)`;

The final goal is to delete the class RandomAccessFile defined
in env.h and leave only the class io::RandomAccessFile defined
in io/random_access_file.h. To achieve this, the first step is
making those two classes have similar APIs, and the next step
would be replacing the usage of the first class with the second one.

stdpain committed:
Fix collect couldn't DB in View (#3864)

HangyuanLiu committed:
Fix compile error in #3864 (#3871)

meegoo commented on be/src/exec/vectorized/csv_scanner.cpp:
```diff
@@ -30,24 +30,29 @@ Status CSVScanner::ScannerCSVReader::_fill_buffer() {
         // Has reached the end of file and the buffer is empty.
         return Status::EndOfFile(_file->filename());
-    } else if (s.size == 0 && _buff.position()[n - 1] != _record_delimiter) {
+    } else if (s.size == 0 && _buff.find(_record_delimiter, n - _record_delimiter.size()) == nullptr) {
```
n could be less than _record_delimiter.size()

chaoyli committed:
BugFix: fix the wrong checksum between different replicas (#3516)

StarRocks uses the CRC algorithm to get the checksums to check the consistency between replicas.
There are two problems. Firstly, the size of chunks between different replicas is not the same.
Secondly, the order of data is not the same because of the out-of-step compaction procedures.

To solve the problem, I change the checksum algorithm to XOR. The result will be stored as a 64-bit integer.
Since the data between replicas is not the same, the checksum will reflect this problem, and the error probability is 1/(2^64).
XOR also obeys the commutative law. I can calculate the checksum Rowset by Rowset. XOR will improve performance and reduce memory consumption. 
    
Using the lineorder table(TPC-H), I test it with 100 million rows and 20 rowsets.
The test indicates that the memory consumption reduced from 500MB to 140MB. And the time is reduced from 24s to 9s.
As the number of rowsets grows, the performance improvement effect will be more obvious.

Youngwb committed:
Disable compute extra cost when can not get accurate table row count (#3818)

murphyatwork committed:
remove deadcode in type_infra (#3876)

trueeyu committed:
format support .tpp file (#3880)

clang format support .tpp file

liuyehcf committed:
Support truncate decimal (#3744)

Youngwb committed:
Fix can not extract plan with single tablet (#3855)

sduzh committed:
Fix compile problem (#3879)

Fix compile problem

ZiheLiu committed:
Fix: decrease memory usage of Analytor (#3848)

sduzh committed:
Merge io::SeekableInputStream and io::RandomAccessFile into one class (#3874)

sevev committed:
Refactor MaterializeTypeConverter (#3799)

Refactor MaterializeTypeConverter
1. Remove duplicate codes to make code easier to read
2. Modify parameter types to meet code specifications
3. Delete some useless log in schema change

stdpain committed:
some refactor and change in UDWF (#3837)

trueeyu committed:
one small refactoring of hash map using in join (#3884)

small refactoring prepared for add DirectingMapHashJoin

* only SerializeHashMap need to create BuildPool and ProbePool
* fix_key_column only need to create once before probe
* some vector for probe need only resize once before all probe

satanson committed:
Make FunctionContext::set_error/has_error/error_msg thread-safe in pipeline engine (#3883)

dirtysalt committed:
Code refactor on `HdfsScanNode` (#3862)

The main purpose of this PR is trying to decouple HdfsScanner & HdfsScanNode, on behalf of further working on porting hdfs scan node to pipeline engine.

Astralidea committed:
Opt.
